iphone/cli/hooks/storyboard.js

```diff
@@ -0,0 +1,591 @@
+	 *
+	 * @access public
+	 */
+	initialize() {
+		this._cli.on('build.pre.compile', {
+			priority: 1200,
+			post: (builder, callback) => {
+				this._logger.trace('Starting storyboard detection');
+				this._builder = builder;
+				this.detectStoryboards().then(callback, e => {
+					if (this._cli.argv.platform === 'ios') {
+						this._logger.error(e.stack);
+					}
+					callback(e);
+				});
+			}
+		});
+
+		this._cli.on('build.ios.xcodeproject', {
+			pre: this.addStoryboardsToXcodeProject.bind(this)
+		});
+	}
+
+	/**
+	 * Detects all available storyboards.
+	 *
+	 * @return {Promise<void>}
+	 * @access private
+	 */
+	async detectStoryboards() {
+		const storyboardPaths = await this.findStoryboardPaths();
+		if (storyboardPaths.length === 0) {
+			return;
+		}
+
+		const incrementalDirectory = path.join(this._builder.projectDir, 'build', 'incrementalStoryboards');
+		const outputDirectory = path.join(this._builder.projectDir, 'build', 'inspectStoryboards');
+		const task = new InspectStoryboardsTask({
+			name: 'ti:inspectStoryboards',
+			logger: this._logger,
+			incrementalDirectory
+		});
+		task.outputDirectory = outputDirectory;
+		storyboardPaths.forEach(storyboardPath => {
+			task.addStoryboardPath(storyboardPath);
+		});
+		task.postTaskRun = () => {
+			this._storyboards = task.storyboards;
+
+			// Convert the internal ES6 map to an object to avoid ES6 in the builder
+			const storyboardObject = {};
+			this._storyboards.forEach(storyboardInfo => {
+				storyboardObject[storyboardInfo.name] = {
+					name: storyboardInfo.name,
+					path: storyboardInfo.path
+				};
+			});
+			this._builder.storyboards = storyboardObject;
+		};
+		return task.run();
+	}
+
+	/**
+	 * Finds any .storyboard file inside modules.
+	 *
+	 * @return {Promise<string[]>} Promise resolving to array of available storyboard paths
+	 * @access private
+	 */
+	async findStoryboardPaths() {
+		const pathsToScan = [];
+		for (let module of this._builder.modules) {
+			pathsToScan.push(path.join(module.modulePath, 'platform'));
+			pathsToScan.push(path.join(module.modulePath, 'Resources'));
+		}
+
+		const results = await Promise.all(pathsToScan.map(pathToScan => this.scanPathForStoryboards(pathToScan)));
+		return results.filter(foundPath => foundPath).reduce((acc, value) => acc.concat(value), []);
+	}
+
+	/**
+	 * Scans the given path for any .storyboard file.
+	 *
+	 * @param {String} storyboardsPath - Path to scan for storybo
+	 * @return {Promise<string[]>}
+	 * @access private
+	 */
+	async scanPathForStoryboards(storyboardsPath) {
+		let files = [];
+		try {
+			files = await fs.readdir(storyboardsPath);
+		} catch (err) {
+			// ignore non-existing directories
+			return;
+		}
+
+		this._logger.trace(`Scanning ${storyboardsPath.cyan} for storyboards`);
+		const foundStoryboardPaths = [];
+		for (const filename of files) {
+			const possibleStoryPath = path.join(storyboardsPath, filename);
```

Comment:
Technically you could delay constructing the full path here to inside the if body and have the `if` just check the filename ends with storyboard. Not sure it makes a huge perf difference...